File: leetcode/algorithms/linked-list/remove-duplicates-from-sorted-list-ii.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Remove Duplicates From Sorted List - II"
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if(head==null){
            return null;
        }
        ListNode cleanedListHead = null, cleanedListTail=null;
        ListNode walkerNode = head;
        
        while(walkerNode!=null){
            ListNode originalWalkerNode = walkerNode;
            while(walkerNode.next!=null && walkerNode.next.val==walkerNode.val){
                walkerNode=walkerNode.next;
            }
            if(walkerNode.equals(originalWalkerNode)){
                ListNode nodeToInsert = new ListNode(walkerNode.val);
                if(cleanedListHead==null){
                    cleanedListHead=nodeToInsert;
                    cleanedListTail=nodeToInsert;
                } else {
                    cleanedListTail.next=nodeToInsert;
                    cleanedListTail=nodeToInsert;
                }
            }
            walkerNode=walkerNode.next;
        }
        
        return cleanedListHead;
        
    }
}

File: leetcode/algorithms/linked-list/remove-duplicates-from-sorted-list-ii.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function deleteDuplicates(head: ListNode | null): ListNode | null {
    if(head===null){
        return head;
    }
    
    let duplicateNumber = null;
    while(head.next!==null && head.next.val===head.val){
        duplicateNumber = head.val;
        head.next = head.next.next;
    }
    
    if(head.val===duplicateNumber){
        return deleteDuplicates(head.next)
    }
    
    head.next = deleteDuplicates(head.next);
    
    
    return head;

};